File: native_plugin/ios/Classes/RetroCamPlugin.swift
import Flutter
import UIKit

public class RetroCamPlugin: NSObject, FlutterPlugin {
    
    private var cameraManager: CameraSessionManager?
    private var renderer: MetalRenderer?
    private var eventSink: FlutterEventSink?
    private var textureRegistry: FlutterTextureRegistry?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.retrocam.app/camera_control", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "com.retrocam.app/camera_events", binaryMessenger: registrar.messenger())
        
        let instance = RetroCamPlugin()
        instance.textureRegistry = registrar.textures()
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initCamera":
            handleInitCamera(call: call, result: result)
        case "startPreview":
            cameraManager?.startSession()
            result(["success": true])
        case "stopPreview":
            cameraManager?.stopSession()
            result(["success": true])
        case "setPreset":
            handleSetPreset(call: call, result: result)
        case "updateLensParams":
            handleUpdateLensParams(call: call, result: result)
        case "takePhoto":
            handleTakePhoto(call: call, result: result)
        case "startRecording":
            handleStartRecording(call: call, result: result)
        case "stopRecording":
            handleStopRecording(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleInitCamera(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // 初始化 CameraManager 和 MetalRenderer
        cameraManager = CameraSessionManager()
        renderer = MetalRenderer(registry: textureRegistry!)
        
        // Register texture with Flutter
        let textureId = textureRegistry!.register(renderer!)
        renderer?.setTextureId(textureId)
        
        cameraManager?.sampleBufferDelegate = renderer!
        cameraManager?.configure()
        
        result(["textureId": textureId])
    }
    
    private func handleSetPreset(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let presetJson = args["preset"] as? [String: Any] else {
            result(FlutterError(code: "1003", message: "Invalid preset parameters", details: nil))
            return
        }
        
        // 优先读取 defaultLook 平铺参数（updateRenderParams 通道传入的完整 shader params）
        var shaderParams: [String: Any] = [:]

        if let defaultLook = presetJson["defaultLook"] as? [String: Any] {
            // 直接将 defaultLook 中所有 NSNumber 字段转为 Float 并写入 shaderParams
            for (key, value) in defaultLook {
                if let num = value as? NSNumber {
                    shaderParams[key] = num.floatValue
                } else if let dict = value as? [String: Any] {
                    shaderParams[key] = dict  // 保留 colorBias 等嵌套字典
                }
            }
        } else {
            // 当传入的是原始 preset JSON 结构时，解析 baseModel
            if let baseModel = presetJson["baseModel"] as? [String: Any] {
                if let color = baseModel["color"] as? [String: Any] {
                    if let contrast   = color["contrast"]   as? NSNumber { shaderParams["contrast"]   = contrast.floatValue }
                    if let saturation = color["saturation"] as? NSNumber { shaderParams["saturation"] = saturation.floatValue }
                }
                if let sensor = baseModel["sensor"] as? [String: Any] {
                    if let noise = sensor["noise"] as? NSNumber { shaderParams["noise"] = noise.floatValue }
                }
            }
        }
        
        renderer?.updateParams(shaderParams)
        
        result(["success": true])
    }
    
    private func handleUpdateLensParams(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // 镜头切换时实时更新 Metal shader 中的镜头参数
        // distortion: Brown-Conrady k1 系数，负值=桶形畸变（鱼眼/广角）
        // vignette: 镜头层暗角，叠加在 preset 暗角之上
        // zoomFactor: UV 缩放，<1.0 时视野扩大（鱼眼圆形遮罩效果）
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "1004", message: "Invalid lens parameters", details: nil))
            return
        }
        let distortion  = (args["distortion"]  as? NSNumber)?.floatValue ?? 0.0
        let vignette    = (args["vignette"]    as? NSNumber)?.floatValue ?? 0.0
        let zoomFactor  = (args["zoomFactor"]  as? NSNumber)?.floatValue ?? 1.0
        renderer?.updateParams([
            "distortion":   distortion,
            "lensVignette": vignette,
            "zoomFactor":   zoomFactor,
        ])
        result(["success": true])
    }
    
    private func handleTakePhoto(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // 触发拍照逻辑，获取高分辨率图像后进行渲染并保存
        // In a full implementation, this would call cameraManager.capturePhoto,
        // wait for the high-res buffer, pass it through the MetalRenderer with the
        // current params, apply paper/border if needed, and save to Photos album.
        
        // Mock implementation for Phase 2 skeleton
        DispatchQueue.global(qos: .userInitiated).async {
            // Simulate processing time
            Thread.sleep(forTimeInterval: 0.5)
            
            DispatchQueue.main.async {
                result(["filePath": "/dummy/path/to/rendered_photo.jpg"])
            }
        }
    }
    
    private func handleStartRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // In a full implementation, this would setup AVAssetWriter, connect it to the
        // MetalRenderer output, and start writing frames to a temporary file.
        result(["success": true])
    }
    
    private func handleStopRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // In a full implementation, this would finish writing the AVAssetWriter,
        // save the resulting MP4 file to the Photos album, and return the path.
        
        DispatchQueue.global(qos: .userInitiated).async {
            Thread.sleep(forTimeInterval: 1.0)
            
            DispatchQueue.main.async {
                result(["filePath": "/dummy/path/to/rendered_video.mp4"])
            }
        }
    }
}

extension RetroCamPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: native_plugin/android/src/main/kotlin/com/retrocam/app/RetroCamPlugin.kt
package com.retrocam.app

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import io.flutter.view.TextureRegistry

class RetroCamPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel : MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    
    private var cameraManager: com.retrocam.app.managers.CameraManager? = null
    private var glRenderer: com.retrocam.app.renderers.GLRenderer? = null
    private lateinit var textureRegistry: TextureRegistry

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        textureRegistry = flutterPluginBinding.textureRegistry
        
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.retrocam.app/camera_control")
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.retrocam.app/camera_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initCamera" -> handleInitCamera(call, result)
            "startPreview" -> {
                cameraManager?.startPreview()
                result.success(mapOf("success" to true))
            }
            "stopPreview" -> {
                cameraManager?.stopPreview()
                result.success(mapOf("success" to true))
            }
            "setPreset" -> handleSetPreset(call, result)
            "takePhoto" -> handleTakePhoto(call, result)
            "startRecording" -> handleStartRecording(call, result)
            "stopRecording" -> handleStopRecording(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleInitCamera(call: MethodCall, result: Result) {
        // 初始化 OpenGL 渲染器并获取 SurfaceTexture
        val textureEntry = textureRegistry.createSurfaceTexture()
        glRenderer = com.retrocam.app.renderers.GLRenderer(textureEntry.surfaceTexture())
        
        // 初始化 CameraX
        // For Flutter plugin, we don't have direct access to a LifecycleOwner in the plugin itself
        // Typically we would use FlutterFragmentActivity or similar
        // For this skeleton, we'll assume cameraManager is initialized correctly later
        
        result.success(mapOf("textureId" to textureEntry.id()))
    }

    private fun handleSetPreset(call: MethodCall, result: Result) {
        val presetJson = call.argument<Map<String, Any>>("preset")
        if (presetJson == null) {
            result.error("1003", "Invalid preset parameters", null)
            return
        }
        
        // Parse preset JSON to extract shader parameters
        val shaderParams = mutableMapOf<String, Any>()
        
        val baseModel = presetJson["baseModel"] as? Map<*, *>
        if (baseModel != null) {
            val color = baseModel["color"] as? Map<*, *>
            if (color != null) {
                color["contrast"]?.let { shaderParams["contrast"] = it }
                color["saturation"]?.let { shaderParams["saturation"] = it }
            }
            val sensor = baseModel["sensor"] as? Map<*, *>
            if (sensor != null) {
                sensor["noise"]?.let { shaderParams["noise"] = it }
            }
        }
        
        glRenderer?.updateParams(shaderParams)
        
        result.success(mapOf("success" to true))
    }

    private fun handleTakePhoto(call: MethodCall, result: Result) {
        // 触发 CameraX 拍照，送入 GL 渲染后保存
        // In a full implementation, this would call cameraManager.takePhoto,
        // wait for the high-res buffer, pass it through the GLRenderer with the
        // current params, apply paper/border if needed, and save to MediaStore.
        
        // Mock implementation for Phase 2 skeleton
        Thread {
            // Simulate processing time
            Thread.sleep(500)
            
            // Return to main thread
            android.os.Handler(android.os.Looper.getMainLooper()).post {
                result.success(mapOf("filePath" to "/dummy/path/to/rendered_photo.jpg"))
            }
        }.start()
    }

    private fun handleStartRecording(call: MethodCall, result: Result) {
        // In a full implementation, this would setup MediaCodec/MediaMuxer,
        // connect it to the GLRenderer output surface, and start writing.
        result.success(mapOf("success" to true))
    }

    private fun handleStopRecording(call: MethodCall, result: Result) {
        // In a full implementation, this would stop the MediaCodec/MediaMuxer,
        // save the resulting MP4 file to MediaStore, and return the path.
        
        Thread {
            Thread.sleep(1000)
            
            android.os.Handler(android.os.Looper.getMainLooper()).post {
                result.success(mapOf("filePath" to "/dummy/path/to/rendered_video.mp4"))
            }
        }.start()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
